rootfs/etc/nginx/template/nginx.tmpl

```diff
@@ -606,7 +606,25 @@ http {
     {{ end }}
 
     {{ range $server := $servers }}
+    {{ range $location := $server.Locations }}
+    {{ $applyGlobalAuth := shouldApplyGlobalAuth $location $all.Cfg.GlobalExternalAuth.URL }}
+    {{ $applyAuthUpstream := shouldApplyAuthUpstream $location $all.Cfg }}
+    {{ if and (eq $applyAuthUpstream true) (eq $applyGlobalAuth false) }}
+    ## start auth upstream {{ $server.Hostname }}{{ $location.Path }}
+    upstream {{ buildAuthUpstreamName $location $server.Hostname }} {
```

Comment:
here I have a question: do we already do this upstreaming via Lua? (I'm not sure. @ElvinEfendi may know).

If yes, adding this may lead us to constant reloads?

Reply:
@rikatz Thank you for the review, it was really helpful!

Answer: For auth requests, the original solution used auth URL directly with `proxy_pass` directive. The downside is that every authentication request opens a new TCP connection creating many TIME_WAITs on the upstream backend. Nginx supports keepalive connections in the `upstream` block and my first solution tried to use that. Unfortunately, Nginx ignores `keepalive` in the `upstream` block with `auth_request` module, that is why we use a similar Lua block instead. While this seems to be working, the Lua `ngx.location.capture` subrequests do not support HTTP/2.

All-in-all we only create a new upstream block and call it from an `access_by_lua_block` (instead of auth_request directive) if `auth-keeaplive` annotation is used, and `use-http2` is disabled.

Reply:
In addition, ingress-nginx already uses `upstream` block with Lua `balancer` module, but external authentication is special because `auth-url` can point to a service running outside of the Kubernetes cluster. This is the reason we create a static upstream block. I don't think this will lead to constant reloads.